enhancements/operator-dependency-resolution.md

```diff
@@ -0,0 +1,216 @@
+
+#### Library
+
+The resolver will be a library that can be imported not just by OLM and used at runtime. This will be a separate component that can be used to determine how OLM will resolve a dependency outside the scope of a running cluster if need be. As a result, we will create a separate git repository for the resolver and have OLM import it.
+
+### Dependency Specification
+
+The dependencies of an operator are listed as a list in `dependencies.yaml` file inside `/metadata` folder of a bundle. This file is optional and only used to specify explicit operator version dependencies at first. Eventually, operator authors can migrate the API-based dependencies into `dependencies.yaml` as well in the future. The ultimate goal is to have `dependencies.yaml` as a centralized metadata for operator dependencies and moving the dependency information away from CSV.
+
+The registry will parse the dependency information in `dependencies.yaml` and add to the database when the bundle is loaded into the registry. OLM will query the dependency information from the registry to facilitate the dependency resolution using sat solver.
+
+### Syntax
+
+The dependency list will contain a `type` field for each item to specify what kind of dependency this is. It can be a package type (`olm.package`) meaning this is a dependency for a specific operator. For `olm.package` type, the dependency information should include the `package` name and the `version` of the package in semver format. We use `blang/semver` library for semver parsing (https://github.com/blang/semver). For example, you can specify an exact version such as `0.5.2` or a range of version such as `>0.5.1` (https://github.com/blang/semver#ranges). In addition, the author can specify dependency that is similiar to existing CRD/API-based using proposed `crd` or `api` type and then specify GVK information as how it is done in CSV. This is a path to enable operator authors to consolidate all dependencies (API or explicit) to be in the same place.
+
+Moreover, for explicit operator dependency, a new field named `indexNamespace` is proposed to enable operator authors to choose a specific source where the operator is provided. This concept is similiar to choosing a specific CatalogSource. However, we choose not to include CatalogSource information in the bundle as that information is out of scope of a bundle. Operator authors shouldn't need to know about all CatalogSources that are available in the cluster as they may not be the same across different clusters. By using this `indexNamespace` type, we require changes in pipeline that build CatalogSources to add a new label `indexNamespace` that indicates the "index namespace" in term of provider where all of operators inside belong to. This `indexNamespace` constrain will allow OLM to pick the correct operator when the same operator is included in different CatalogSource.
```

Comment:
I don't think we need pipeline changes for this, do we?

We also discussed having `indexNamespace` set per package, maybe a discussion here of why that might be useful?

Reply:
My memory is falling me on this part a bit but I recall something related to pipeline about `indexNamespace` part. If we decide to specify this `indexNamespace` per package, I wonder where this label should be included. Should the label be included in the bundle hence the author will need to choose which indexNamespace it should be in?

I kind of envision pipeline would pack a bunch of operators in the same indexNamespace upon the choice of pipeline team. Does that make sense?

Reply:
Didn't we have some discussion about concatenating namespaces?

i.e. a pipeline wants to build the "etcd" index which would have its own namespace. Then that gets merged into the "redhat-operators" index. Would the namespace just be overridden? Are they concatenated into a new namespace? Could it be a list of registered namespaces?

Reply:
I don't recall we had a convo about concatenating namespaces or maybe I just forgot it.

I want to argue that `indexNamespace` should be something that pipeline should care about as they build index image. It is essentially a way for them to group bundle together. I don't see the reason why authors need to specify `indexNamespace` for the bundles.

I think when you build an index image with an `indexNamespace` label, then it is immutable in that image unless you rebuild the image and change the label. I don't expect cluster-admin to update the label but instead build a new image with different label.

At the moment, I think it is possible to select dependencies if they are multiple choices in `InstallPlan` after dependency resolution. However, after a choice is made, we probably need to resolve it again to ensure we pick up other dependencies that the chosen operator may need. Would that be too much?

I simply see of there are two indexes with same `indexNamespace` then they are merged together as a single index with all operators inside under the same `indexNamespace` in the perspective of dependency resolution.